File: backend/app/schemas/utility_kpi.py
"""
Utility KPI Center — Pydantic v2 Schemas
All KPI payloads served by the /kpi endpoints.
"""
from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel


# ── Shared ────────────────────────────────────────────────────────────────────

class TopConsumerRow(BaseModel):
    label:       str
    consumption: Decimal
    unit:        str
    cost:        Optional[Decimal] = None
    pct_of_total: Optional[Decimal] = None
    record_count: int = 0


class TopConsumers(BaseModel):
    dimension:    str          # department | production_line | machine_ref
    utility_type: Optional[str] = None
    date_from:    Optional[str] = None
    date_to:      Optional[str] = None
    rows:         List[TopConsumerRow]
    grand_total:  Decimal


class DailyTrendPoint(BaseModel):
    date:             str
    electricity_kwh:  Optional[Decimal] = None
    water_m3:         Optional[Decimal] = None
    gas_nm3:          Optional[Decimal] = None
    steam_ton:        Optional[Decimal] = None
    solar_kwh:        Optional[Decimal] = None
    compressed_air_nm3: Optional[Decimal] = None
    total_cost:       Optional[Decimal] = None
    is_estimated:     bool = False


# ── Main dashboard ─────────────────────────────────────────────────────────────

class MainDashboardKPIs(BaseModel):
    report_date:   str
    period_label:  str = "Today"

    # Electricity
    electricity_kwh_today:     Optional[Decimal] = None
    electricity_kwh_mtd:       Optional[Decimal] = None
    electricity_kwh_per_ton:   Optional[Decimal] = None
    peak_demand_kw:            Optional[Decimal] = None
    night_base_load_kwh:       Optional[Decimal] = None

    # Water
    water_m3_today:            Optional[Decimal] = None
    water_m3_per_ton:          Optional[Decimal] = None

    # Gas
    gas_nm3_today:             Optional[Decimal] = None
    gas_nm3_per_ton:           Optional[Decimal] = None

    # Steam / Boiler
    steam_ton_today:           Optional[Decimal] = None
    gas_per_ton_steam:         Optional[Decimal] = None
    condensate_recovery_pct:   Optional[Decimal] = None
    blowdown_loss_pct:         Optional[Decimal] = None
    boiler_efficiency_pct:     Optional[Decimal] = None

    # Compressed Air
    air_nm3_today:             Optional[Decimal] = None
    air_kwh_per_nm3:           Optional[Decimal] = None
    leak_estimate_pct:         Optional[Decimal] = None
    compressor_efficiency_pct: Optional[Decimal] = None

    # Soft Water
    soft_water_m3_today:       Optional[Decimal] = None
    soft_water_conversion_pct: Optional[Decimal] = None
    salt_per_m3_soft:          Optional[Decimal] = None

    # Solar
    solar_kwh_today:           Optional[Decimal] = None
    solar_contribution_pct:    Optional[Decimal] = None
    solar_vs_expected_pct:     Optional[Decimal] = None

    # Wastewater
    wastewater_cost_per_m3:    Optional[Decimal] = None
    cod_removal_pct:           Optional[Decimal] = None
    wastewater_compliance:     Optional[str] = None

    # Treatment Chemicals
    chemical_cost_today:       Optional[Decimal] = None

    # Cost
    utility_cost_today:        Optional[Decimal] = None
    utility_cost_mtd:          Optional[Decimal] = None
    utility_cost_per_ton:      Optional[Decimal] = None
    utility_cost_per_batch:    Optional[Decimal] = None
    currency_code:             str = "USD"

    # Machine utility
    machine_standby_pct:       Optional[Decimal] = None
    dept_utility_intensity:    Optional[Decimal] = None   # cost/m² or cost/ton per dept

    # Production reference (supplied externally or derived)
    production_tons_today:     Optional[Decimal] = None

    # Alarms
    critical_alarm_count: int = 0
    warning_alarm_count:  int = 0

    # Data quality flag
    has_estimated_data: bool = False


# ── Per-utility sub-dashboard KPIs ────────────────────────────────────────────

class ElectricityDashboardKPIs(BaseModel):
    date_from:            str
    date_to:              str
    kwh_total:            Optional[Decimal] = None
    kwh_per_ton:          Optional[Decimal] = None
    peak_demand_kw:       Optional[Decimal] = None
    night_base_load_kwh:  Optional[Decimal] = None
    solar_offset_kwh:     Optional[Decimal] = None
    solar_contribution_pct: Optional[Decimal] = None
    total_cost:           Optional[Decimal] = None
    cost_per_kwh:         Optional[Decimal] = None
    estimated_pct:        Optional[Decimal] = None
    anomaly_count:        int = 0


class WaterDashboardKPIs(BaseModel):
    date_from:      str
    date_to:        str
    water_m3_total: Optional[Decimal] = None
    water_m3_per_ton: Optional[Decimal] = None
    process_water_m3: Optional[Decimal] = None
    total_cost:     Optional[Decimal] = None
    cost_per_m3:    Optional[Decimal] = None
    estimated_pct:  Optional[Decimal] = None
    anomaly_count:  int = 0


class BoilerDashboardKPIs(BaseModel):
    date_from:              str
    date_to:                str
    steam_ton_total:        Optional[Decimal] = None
    fuel_consumed_total:    Optional[Decimal] = None
    fuel_unit:              Optional[str] = None
    gas_per_ton_steam:      Optional[Decimal] = None
    boiler_efficiency_pct:  Optional[Decimal] = None
    condensate_recovery_pct: Optional[Decimal] = None
    blowdown_loss_pct:      Optional[Decimal] = None
    feedwater_m3_total:     Optional[Decimal] = None
    burner_runtime_hours:   Optional[Decimal] = None
    avg_boiler_load_pct:    Optional[Decimal] = None
    anomaly_count:          int = 0


class CompressorDashboardKPIs(BaseModel):
    date_from:              str
    date_to:                str
    air_nm3_total:          Optional[Decimal] = None
    electricity_kwh_total:  Optional[Decimal] = None
    air_kwh_per_nm3:        Optional[Decimal] = None
    leak_estimate_pct:      Optional[Decimal] = None
    night_idle_kwh:         Optional[Decimal] = None
    avg_load_pct:           Optional[Decimal] = None
    runtime_hours:          Optional[Decimal] = None
    anomaly_count:          int = 0


class SoftWaterDashboardKPIs(BaseModel):
    date_from:              str
    date_to:                str
    sw_volume_m3:           Optional[Decimal] = None
    raw_input_m3:           Optional[Decimal] = None
    conversion_efficiency_pct: Optional[Decimal] = None
    salt_consumed_kg:       Optional[Decimal] = None
    salt_per_m3:            Optional[Decimal] = None
    avg_product_hardness:   Optional[Decimal] = None
    regen_count:            Optional[int] = None
    anomaly_count:          int = 0


class SolarDashboardKPIs(BaseModel):
    date_from:              str
    date_to:                str
    kwh_generated:          Optional[Decimal] = None
    kwh_self_consumed:      Optional[Decimal] = None
    kwh_exported:           Optional[Decimal] = None
    solar_contribution_pct: Optional[Decimal] = None
    avg_pr_ratio:           Optional[Decimal] = None
    avg_capacity_factor:    Optional[Decimal] = None
    avg_inverter_efficiency: Optional[Decimal] = None
    anomaly_count:          int = 0


class ChemicalDashboardKPIs(BaseModel):
    date_from:          str
    date_to:            str
    total_cost:         Optional[Decimal] = None
    total_consumed_kg:  Optional[Decimal] = None
    cost_per_day:       Optional[Decimal] = None
    distinct_chemicals: int = 0
    boiler_cost:        Optional[Decimal] = None
    cooling_cost:       Optional[Decimal] = None
    wwtp_cost:          Optional[Decimal] = None


class WastewaterDashboardKPIs(BaseModel):
    date_from:          str
    date_to:            str
    cod_removal_pct:    Optional[Decimal] = None
    bod_removal_pct:    Optional[Decimal] = None
    avg_do_mgl:         Optional[Decimal] = None
    avg_effluent_ph:    Optional[Decimal] = None
    sludge_kg_total:    Optional[Decimal] = None
    power_kwh_total:    Optional[Decimal] = None
    power_per_m3:       Optional[Decimal] = None
    compliant_pct:      Optional[Decimal] = None
    non_compliant_count: int = 0
    anomaly_count:      int = 0


class MachineDashboardKPIs(BaseModel):
    date_from:              str
    date_to:                str
    total_consumption:      Optional[Decimal] = None
    consumption_unit:       Optional[str] = None
    electricity_per_runtime_h: Optional[Decimal] = None
    water_per_batch:        Optional[Decimal] = None
    air_per_batch:          Optional[Decimal] = None
    standby_pct:            Optional[Decimal] = None
    cost_per_batch:         Optional[Decimal] = None
    avg_variance_pct:       Optional[Decimal] = None
    anomaly_count:          int = 0
    flag_high_variance:     bool = False
    flag_standby_waste:     bool = False


class UtilityCostDashboardKPIs(BaseModel):
    date_from:          str
    date_to:            str
    total_cost:         Optional[Decimal] = None
    electricity_cost:   Optional[Decimal] = None
    water_cost:         Optional[Decimal] = None
    gas_cost:           Optional[Decimal] = None
    steam_cost:         Optional[Decimal] = None
    other_cost:         Optional[Decimal] = None
    cost_per_ton:       Optional[Decimal] = None
    cost_per_batch:     Optional[Decimal] = None
    total_bills:        int = 0
    unpaid_amount:      Optional[Decimal] = None
    currency_code:      str = "USD"

File: frontend/src/lib/utilityKpi.ts
/**
 * Utility KPI Center — TypeScript types and API client.
 * All KPI formulas are computed server-side; this file only declares shapes.
 */
import { apiClient } from "./api";

// ── Shared types ──────────────────────────────────────────────────────────────

export interface TopConsumerRow {
  label:        string;
  consumption:  number;
  unit:         string;
  cost?:        number | null;
  pct_of_total?: number | null;
  record_count: number;
}

export interface TopConsumers {
  dimension:    string;
  utility_type?: string | null;
  date_from?:   string | null;
  date_to?:     string | null;
  rows:         TopConsumerRow[];
  grand_total:  number;
}

export interface DailyTrendPoint {
  date:              string;
  electricity_kwh?:  number | null;
  water_m3?:         number | null;
  gas_nm3?:          number | null;
  steam_ton?:        number | null;
  solar_kwh?:        number | null;
  compressed_air_nm3?: number | null;
  total_cost?:       number | null;
  is_estimated:      boolean;
}

// ── Main dashboard KPIs ───────────────────────────────────────────────────────

export interface MainDashboardKPIs {
  report_date:   string;
  period_label:  string;

  electricity_kwh_today?:     number | null;
  electricity_kwh_mtd?:       number | null;
  electricity_kwh_per_ton?:   number | null;
  peak_demand_kw?:            number | null;
  night_base_load_kwh?:       number | null;

  water_m3_today?:            number | null;
  water_m3_per_ton?:          number | null;

  gas_nm3_today?:             number | null;
  gas_nm3_per_ton?:           number | null;

  steam_ton_today?:           number | null;
  gas_per_ton_steam?:         number | null;
  condensate_recovery_pct?:   number | null;
  blowdown_loss_pct?:         number | null;
  boiler_efficiency_pct?:     number | null;

  air_nm3_today?:             number | null;
  air_kwh_per_nm3?:           number | null;
  leak_estimate_pct?:         number | null;
  compressor_efficiency_pct?: number | null;

  soft_water_m3_today?:       number | null;
  soft_water_conversion_pct?: number | null;
  salt_per_m3_soft?:          number | null;

  solar_kwh_today?:           number | null;
  solar_contribution_pct?:    number | null;
  solar_vs_expected_pct?:     number | null;

  wastewater_cost_per_m3?:    number | null;
  cod_removal_pct?:           number | null;
  wastewater_compliance?:     string | null;

  chemical_cost_today?:       number | null;

  utility_cost_today?:        number | null;
  utility_cost_mtd?:          number | null;
  utility_cost_per_ton?:      number | null;
  utility_cost_per_batch?:    number | null;
  currency_code:              string;

  machine_standby_pct?:       number | null;
  dept_utility_intensity?:    number | null;

  production_tons_today?:     number | null;
  critical_alarm_count:       number;
  warning_alarm_count:        number;
  has_estimated_data:         boolean;
}

// ── Per-utility KPI types ─────────────────────────────────────────────────────

export interface ElectricityDashboardKPIs {
  date_from: string; date_to: string;
  kwh_total?:            number | null;
  kwh_per_ton?:          number | null;
  peak_demand_kw?:       number | null;
  night_base_load_kwh?:  number | null;
  solar_offset_kwh?:     number | null;
  solar_contribution_pct?: number | null;
  total_cost?:           number | null;
  cost_per_kwh?:         number | null;
  estimated_pct?:        number | null;
  anomaly_count:         number;
}

export interface WaterDashboardKPIs {
  date_from: string; date_to: string;
  water_m3_total?:  number | null;
  water_m3_per_ton?: number | null;
  process_water_m3?: number | null;
  total_cost?:       number | null;
  cost_per_m3?:      number | null;
  estimated_pct?:    number | null;
  anomaly_count:     number;
}

export interface BoilerDashboardKPIs {
  date_from: string; date_to: string;
  steam_ton_total?:         number | null;
  fuel_consumed_total?:     number | null;
  fuel_unit?:               string | null;
  gas_per_ton_steam?:       number | null;
  boiler_efficiency_pct?:   number | null;
  condensate_recovery_pct?: number | null;
  blowdown_loss_pct?:       number | null;
  feedwater_m3_total?:      number | null;
  burner_runtime_hours?:    number | null;
  avg_boiler_load_pct?:     number | null;
  anomaly_count:            number;
}

export interface CompressorDashboardKPIs {
  date_from: string; date_to: string;
  air_nm3_total?:         number | null;
  electricity_kwh_total?: number | null;
  air_kwh_per_nm3?:       number | null;
  leak_estimate_pct?:     number | null;
  night_idle_kwh?:        number | null;
  avg_load_pct?:          number | null;
  runtime_hours?:         number | null;
  anomaly_count:          number;
}

export interface SoftWaterDashboardKPIs {
  date_from: string; date_to: string;
  sw_volume_m3?:             number | null;
  raw_input_m3?:             number | null;
  conversion_efficiency_pct?: number | null;
  salt_consumed_kg?:         number | null;
  salt_per_m3?:              number | null;
  avg_product_hardness?:     number | null;
  regen_count?:              number | null;
  anomaly_count:             number;
}

export interface SolarDashboardKPIs {
  date_from: string; date_to: string;
  kwh_generated?:          number | null;
  kwh_self_consumed?:      number | null;
  kwh_exported?:           number | null;
  solar_contribution_pct?: number | null;
  avg_pr_ratio?:           number | null;
  avg_capacity_factor?:    number | null;
  avg_inverter_efficiency?: number | null;
  anomaly_count:           number;
}

export interface ChemicalDashboardKPIs {
  date_from: string; date_to: string;
  total_cost?:        number | null;
  total_consumed_kg?: number | null;
  cost_per_day?:      number | null;
  distinct_chemicals: number;
  boiler_cost?:       number | null;
  cooling_cost?:      number | null;
  wwtp_cost?:         number | null;
}

export interface WastewaterDashboardKPIs {
  date_from: string; date_to: string;
  cod_removal_pct?:     number | null;
  bod_removal_pct?:     number | null;
  avg_do_mgl?:          number | null;
  avg_effluent_ph?:     number | null;
  sludge_kg_total?:     number | null;
  power_kwh_total?:     number | null;
  power_per_m3?:        number | null;
  compliant_pct?:       number | null;
  non_compliant_count:  number;
  anomaly_count:        number;
}

export interface MachineDashboardKPIs {
  date_from: string; date_to: string;
  total_consumption?:       number | null;
  consumption_unit?:        string | null;
  electricity_per_runtime_h?: number | null;
  water_per_batch?:         number | null;
  air_per_batch?:           number | null;
  standby_pct?:             number | null;
  cost_per_batch?:          number | null;
  avg_variance_pct?:        number | null;
  anomaly_count:            number;
  flag_high_variance:       boolean;
  flag_standby_waste:       boolean;
}

export interface UtilityCostDashboardKPIs {
  date_from: string; date_to: string;
  total_cost?:      number | null;
  electricity_cost?: number | null;
  water_cost?:      number | null;
  gas_cost?:        number | null;
  steam_cost?:      number | null;
  other_cost?:      number | null;
  cost_per_ton?:    number | null;
  cost_per_batch?:  number | null;
  total_bills:      number;
  unpaid_amount?:   number | null;
  currency_code:    string;
}

// ── Color / label maps for charts ─────────────────────────────────────────────

export const UTILITY_CHART_COLORS: Record<string, string> = {
  electricity_kwh:     "#fbbf24",
  water_m3:            "#60a5fa",
  gas_nm3:             "#a78bfa",
  steam_ton:           "#f97316",
  solar_kwh:           "#facc15",
  compressed_air_nm3:  "#22d3ee",
  total_cost:          "#94a3b8",
};

export const UTILITY_CHART_LABELS: Record<string, string> = {
  electricity_kwh:    "Electricity (kWh)",
  water_m3:           "Water (m³)",
  gas_nm3:            "Gas (Nm³)",
  steam_ton:          "Steam (t)",
  solar_kwh:          "Solar (kWh)",
  compressed_air_nm3: "Compressed Air (Nm³)",
  total_cost:         "Cost",
};

// ── API client ────────────────────────────────────────────────────────────────

export const kpiApi = {
  // Main dashboard
  async mainDashboard(params?: { report_date?: string; production_tons?: number }): Promise<MainDashboardKPIs> {
    const r = await apiClient.get<MainDashboardKPIs>("/api/v1/kpi/dashboard", { params }).then((r) => r.data);
    return r.data;
  },

  // Top consumers
  async topConsumers(params?: {
    dimension?: string; utility_type?: string;
    date_from?: string; date_to?: string; top_n?: number;
  }): Promise<TopConsumers> {
    const r = await apiClient.get<TopConsumers>("/api/v1/kpi/top-consumers", { params }).then((r) => r.data);
    return r.data;
  },

  // Daily trend
  async dailyTrend(params?: { date_from?: string; date_to?: string }): Promise<DailyTrendPoint[]> {
    const r = await apiClient.get<DailyTrendPoint[]>("/api/v1/kpi/trend", { params }).then((r) => r.data);
    return r.data;
  },

  // Sub-dashboards
  async electricity(params?: { date_from?: string; date_to?: string }): Promise<ElectricityDashboardKPIs> {
    const r = await apiClient.get<ElectricityDashboardKPIs>("/api/v1/kpi/electricity", { params }).then((r) => r.data);
    return r.data;
  },
  async water(params?: { date_from?: string; date_to?: string }): Promise<WaterDashboardKPIs> {
    const r = await apiClient.get<WaterDashboardKPIs>("/api/v1/kpi/water", { params }).then((r) => r.data);
    return r.data;
  },
  async boiler(params?: { date_from?: string; date_to?: string }): Promise<BoilerDashboardKPIs> {
    const r = await apiClient.get<BoilerDashboardKPIs>("/api/v1/kpi/boiler", { params }).then((r) => r.data);
    return r.data;
  },
  async compressor(params?: { date_from?: string; date_to?: string }): Promise<CompressorDashboardKPIs> {
    const r = await apiClient.get<CompressorDashboardKPIs>("/api/v1/kpi/compressor", { params }).then((r) => r.data);
    return r.data;
  },
  async softWater(params?: { date_from?: string; date_to?: string }): Promise<SoftWaterDashboardKPIs> {
    const r = await apiClient.get<SoftWaterDashboardKPIs>("/api/v1/kpi/soft-water", { params }).then((r) => r.data);
    return r.data;
  },
  async solar(params?: { date_from?: string; date_to?: string }): Promise<SolarDashboardKPIs> {
    const r = await apiClient.get<SolarDashboardKPIs>("/api/v1/kpi/solar", { params }).then((r) => r.data);
    return r.data;
  },
  async chemicals(params?: { date_from?: string; date_to?: string }): Promise<ChemicalDashboardKPIs> {
    const r = await apiClient.get<ChemicalDashboardKPIs>("/api/v1/kpi/chemicals", { params }).then((r) => r.data);
    return r.data;
  },
  async wastewater(params?: { date_from?: string; date_to?: string }): Promise<WastewaterDashboardKPIs> {
    const r = await apiClient.get<WastewaterDashboardKPIs>("/api/v1/kpi/wastewater", { params }).then((r) => r.data);
    return r.data;
  },
  async machineUtility(params?: { date_from?: string; date_to?: string }): Promise<MachineDashboardKPIs> {
    const r = await apiClient.get<MachineDashboardKPIs>("/api/v1/kpi/machine-utility", { params }).then((r) => r.data);
    return r.data;
  },
  async utilityCost(params?: { date_from?: string; date_to?: string }): Promise<UtilityCostDashboardKPIs> {
    const r = await apiClient.get<UtilityCostDashboardKPIs>("/api/v1/kpi/utility-cost", { params }).then((r) => r.data);
    return r.data;
  },
};
